### ISSUE TITLE
 Source abilities for continuous effects don't keep zone change counter current

### ISSUE THREAD

**Issue by theelk801** [magefree/mage#8706]
Continuous effects have a map which links them to their source ability, but the map is initialized at the start of the game and the source ability that they're mapped to doesn't change. This means that when applying a continuous effect, the zone change counter of its ability isn't actually current and is stuck at 0. This is different from activated or activated abilities which do stay current. 

There seems to be a workaround for this in `AbilityImpl.getSourceObjectIfItStillExists(game)` which prevents some issues:
https://github.com/magefree/mage/blob/54203c16d3505dcfb55da5f861beea160ad66f17/Mage/src/main/java/mage/abilities/AbilityImpl.java#L1265-L1273

But anything that calls `source.getSourceObjectZoneChangeCounter()`, like for example `CardUtil.getExileZoneId`, will still get the wrong info.

I'd like to fix this but I'm not sure how to do so at the moment. For now, a usable workaround is calling `game.getState().getZoneChangeCounter(source.getSourceId())` instead of `source.getSourceObjectZoneChangeCounter()` inside of a continuous effect, which definitely works when the effect is only in one zone as other separate mechanisms guarantee that the object you get from `game.getObject` will be the source object. I do still think this is fixable though.

**Comment by theelk801** [comment]
my first attempt was this in `MageObjectImpl`:

<img width="671" alt="image" src="https://user-images.githubusercontent.com/29759218/154586996-363ae8e5-326f-4c33-b080-88d00e954200.png">

It caused a ton of test failures but I'm thinking it's a good start

**Comment by JayDi85** [comment]
Nope, you can't sync source's zcc with ability's zcc -- it's a different things. Ability zcc must be static (moment of the init), but object's zcc is dynamic (changes on object's zone changing).

**Comment by JayDi85** [comment]
And I don't remember how's it works with singletone abilities. Weird.

**Comment by JayDi85** [comment]
Looks like it’s not a bug, it’s by design cause some continues effects need to keep starting zcc. Example with sacrifice from https://github.com/magefree/mage/issues/10479#issuecomment-2436074171

![shot_241024_222634](https://github.com/user-attachments/assets/f1343574-ebde-4f5d-809e-af96f6dbd887)
